daemon/containerd/lazy_push.go

```
@@ -0,0 +1,514 @@
+				if desc.Size > 0 {
+					log.WithError(err).Debug("failed to resolve missing content, fallback to original")
+					name = ref.String()
+					resolved = desc
+				} else {
+					return err
+				}
+			}
+
+			log.WithField("name", name).Debug("resolved missing content")
+			fetcher, err := resolver.Fetcher(ctx, name)
+			if err != nil {
+				log.WithError(err).Debug("failed to create fetcher")
+				return err
+			}
+
+			appendDistributionSourceLabel, err := docker.AppendDistributionSourceLabel(store, ref.String())
+			if err != nil {
+				return err
+			}
+
+			fetchHandler := containerdimages.Handlers(
+				remotes.FetchHandler(store, fetcher),
+				appendDistributionSourceLabel,
+				appendLabelHandler(ctx, store, "docker.io/fetch.reason", "push"),
+			)
+
+			_, err = fetchHandler(ctx, resolved)
+			if err != nil {
+				log.WithError(err).Debug("failed to fetch")
+				return err
+			}
+
+			log.Debug("fetched!")
+			mutex.Lock()
+			fetched = append(fetched, desc)
+			mutex.Unlock()
+			return nil
+		})
+	}
+
+	err := eg.Wait()
+	return fetched, err
+}
+
+// appendLabelHandler returns a handler which adds a label with value to the handled content
+func appendLabelHandler(ctx context.Context, manager content.Manager, key, value string) containerdimages.HandlerFunc {
+	return containerdimages.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) (subdescs []ocispec.Descriptor, err error) {
+		info, err := manager.Info(ctx, desc.Digest)
+		if err != nil {
+			return nil, err
+		}
+		if info.Labels == nil {
+			info.Labels = map[string]string{}
+		}
+		info.Labels[key] = value
+
+		_, err = manager.Update(ctx, info, "labels."+key)
+		return nil, err
+	})
+}
+
+// contentDoesntExist returns true only if content is not present in store.
+// If there is other error when querying the content it returns false.
+func contentDoesntExist(ctx context.Context, store content.Store, desc ocispec.Descriptor) bool {
```

I don't think we should swallow the error, maybe it's better to return it?

---

I like the readability of returning just one value so the usage looks like this:
```go
if contentDoesntExist(ctx, store, desc) 
```

If other error is encountered we will get it again anyway after trying to do something with the content, in this case calling `children, err := containerdimages.Children(ctx, store, desc)`

---

Sure, that's the case _now_, but someone else could use this in the future without calling `conatinerdimages.Children` and get a `false` and not know that there was an error

---

Done